src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,516 @@
+	apdu.resplen = len;
+	apdu.le = len;
+
+	r = sc_transmit_apdu(card, &apdu);
+	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	return r;
+}
+
+#define ATR_MATCH 1
+
+static int eoi_match_card(sc_card_t* card) {
+	LOG_FUNC_CALLED(card->ctx);
+	if (_sc_match_atr(card, eoi_atrs, &card->type) >= 0) {
+		sc_log(card->ctx, "ATR recognized as Slovenian eID card");
+		LOG_FUNC_RETURN(card->ctx, ATR_MATCH);
+	}
+	LOG_FUNC_RETURN(card->ctx, !ATR_MATCH);
+}
+
+static int eoi_init(sc_card_t* card) {
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	u8 version[6];
+	size_t i, j;
+	scconf_block **found_blocks, *block;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (eoi_get_data(card, 0x16, version, sizeof(version)) != SC_SUCCESS)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_WRONG_CARD);
+
+	if (privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+	privdata = calloc(1, sizeof(struct eoi_privdata));
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+	card->drv_data = privdata;
+
+	sprintf(privdata->version, "%X%02X.%02X%02X", version[0], version[1], version[2], version[3]);
+	sc_log(card->ctx, "App version: %s", privdata->version);
+
+	memset(&card->sm_ctx, 0, sizeof card->sm_ctx);
+	card->sm_ctx.ops.open = eoi_sm_open;
+
+	card->max_send_size = SC_MAX_APDU_RESP_SIZE;
+	card->max_recv_size = SC_MAX_APDU_RESP_SIZE;
+
+	for (i = 0; i < sizeof eoi_curves / sizeof * eoi_curves; ++i) {
+		LOG_TEST_RET(card->ctx, _sc_card_add_ec_alg(
+			card, eoi_curves[i].len,
+			SC_ALGORITHM_ECDSA_RAW | SC_ALGORITHM_ECDSA_HASH_NONE,
+			0, &eoi_curves[i].oid
+		), "Add EC alg failed");
+	}
+
+	privdata->can = getenv("EOI_CAN");
+	for (i = 0; card->ctx->conf_blocks[i]; i++) {
+		found_blocks = scconf_find_blocks(card->ctx->conf, card->ctx->conf_blocks[i],
+					"card_driver", "eoi");
+		if (!found_blocks)
+			continue;
+
+		for (j = 0, block = found_blocks[j]; block; j++, block = found_blocks[j]) {
+			if (!privdata->can)
+				privdata->can = scconf_get_str(block, "can", NULL);
+		}
+	}
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int eoi_finish(sc_card_t* card)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (privdata) {
+		memset(privdata, 0, sizeof(struct eoi_privdata));
+		free(privdata);
+	}
+
+	card->drv_data = NULL;
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int eoi_select_file(sc_card_t *card, const sc_path_t *in_path, sc_file_t **file_out)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	int r, i;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	r = sc_get_iso7816_driver()->ops->select_file(card, in_path, file_out);
+
+	for (i = 0; i < MAX_OBJECTS && privdata->pin_paths[i]; i++) {
+		if (privdata->pin_paths[i] && sc_compare_path(privdata->pin_paths[i], in_path) && r == SC_ERROR_FILE_NOT_FOUND) {
```

Comment:
If you're checking for `r == SC_ERROR_FILE_NOT_FOUND`, right after calling sc_get_iso7816_driver, then you don't need to loop through all those objects.

Reply:
The problem here is, that in a `use_key()` function a `sc_select_file(PrKey)` is called and on our eID card this call fails (and consequently also complete signing operation). To solve this a failed `sc_select_file()` (for a private key) is intercepted and `SC_ERROR_FILE_NOT_FOUND` replaced with `SC_SUCCESS`. Kind of a hack, but it works.

Reply:
I understand, but my suggestion was to optimize this to something like this:
```
r = sc_get_iso7816_driver()->ops->select_file(card, in_path, file_out);

if (r == SC_ERROR_FILE_NOT_FOUND) {
	for (i = 0; i < MAX_OBJECTS && privdata->pin_paths[i]; i++) {
		if (privdata->pin_paths[i]) {
			...
		}
	}
}
```
And on a second look, I think it would be even preferably to first check the path whether the private key needs to be selected and only call sc_get_iso7816_driver()->ops->select_file if it is not. This would avoid the card error and would look something like this:
```
static int eoi_select_file(sc_card_t *card, const sc_path_t *in_path, sc_file_t **file_out)
{
	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
	int r, i;

	LOG_FUNC_CALLED(card->ctx);

	if (!privdata)
		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);

	for (i = 0; i < MAX_OBJECTS && privdata->pin_paths[i]; i++) {
		if (privdata->pin_paths[i] && sc_compare_path(privdata->pin_paths[i], in_path)) {
			LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
		}
	}

	r = sc_get_iso7816_driver()->ops->select_file(card, in_path, file_out);

	LOG_FUNC_RETURN(card->ctx, r);
}
```

Reply:
Done.